src/main/java/org/threadly/concurrent/future/FutureUtils.java

```diff
@@ -883,6 +883,67 @@ public void handleResult(ST result) {
   }
   
   /**
+   * Similar to {@link #transform(ListenableFuture, Function, Executor)} except designed to handle 
+   * functions which return futures.  This will take what otherwise would be 
+   * {@code ListenableFuture<ListenableFuture<R>>}, and flattens it into a single future which will 
+   * resolve once the contained future is complete.
+   * 
+   * @since 5.0.0
+   * 
+   * @param <ST> The source type for the object returned from the future and inputed into the mapper
+   * @param <RT> The result type for the object contained in the future returned from the mapper
+   * @param sourceFuture Future to source input into transformation function
+   * @param transformFunction Function to apply result from future into returned future
+   * @param executor Executor to execute transformation function on, or {@code null}
+   * @return Future with result of transformation function or respective error
+   */
+  protected static <ST, RT> ListenableFuture<RT> transformAndFlatten(ListenableFuture<ST> sourceFuture, 
+                                                                     Function<? super ST, ListenableFuture<RT>> transformFunction, 
+                                                                     Executor executor) {
+    if (executor == null && sourceFuture.isDone() && ! sourceFuture.isCancelled()) {
+      // optimized path for already complete futures which we can now process in thread
+      try {
+        return transformFunction.apply(sourceFuture.get());
+      } catch (InterruptedException e) {
+        // should not be possible
+        throw new RuntimeException(e);
+      } catch (ExecutionException e) {
+        // failure in getting result from future, transfer failure
+        return FutureUtils.immediateFailureFuture(e.getCause());
+      } catch (Throwable t) {
+        // failure calculating transformation
+        return FutureUtils.immediateFailureFuture(t);
+      }
+    } else {
+      SettableListenableFuture<RT> slf;
+      if (sourceFuture.isCancelled()) {
+        // shortcut
+        slf = new SettableListenableFuture<>();
+        slf.cancel(false);
+        return slf;
+      }
+      slf = new CancelDelegateSettableListenableFuture<>(sourceFuture);
+      sourceFuture.addCallback(new FailurePropogatingFutureCallback<ST>(slf) {
+        @Override
+        public void handleResult(ST result) {
+          try {
+            transformFunction.apply(result)
+                             .addCallback(new FailurePropogatingFutureCallback<RT>(slf) {
+              @Override
+              public void handleResult(RT result) {
+                slf.setResult(result);
+              }
+            });
+          } catch (Throwable t) {
+            slf.setFailure(t);
+          }
+        }
+      }, executor);
+      return slf;
+    }
```

Comment:
> The actual implementation of it looks very close to map, I wanted to try and reduce the code duplication here some, but struggled to find a way to do it without increasing the GC overhead (which for these can already get a bit crazy anyways).

Could you share everything but the immediate done case and the `FailurePropagatingFutureCallback` body? Attempt:

```java
  private static <ST, RT> ListenableFuture<RT> thing(LF<ST> sourceFuture,
                                                     Function<ST, LF<RT>> doneTransform,
                                                     FPFC<ST> cb) {
    if (executor == null && sourceFuture.isDone() && ! sourceFuture.isCancelled()) {
      // optimized path for already complete futures which we can now process in thread
      try {
        return doneTransform.apply(sourceFuture.get());
      } catch (InterruptedException e) {
        // should not be possible
        throw new RuntimeException(e);
      } catch (ExecutionException e) {
        // failure in getting result from future, transfer failure
        return FutureUtils.immediateFailureFuture(e.getCause());
      } catch (Throwable t) {
        // failure calculating transformation
        return FutureUtils.immediateFailureFuture(t);
      }
    } else {
      SettableListenableFuture<RT> slf;
      if (sourceFuture.isCancelled()) {
        // shortcut
        slf = new SettableListenableFuture<>();
        slf.cancel(false);
        return slf;
      }
      slf = new CancelDelegateSettableListenableFuture<>(sourceFuture);
      sourceFuture.addCallback(cb, executor);
      return slf;
    }
  }

 protected static <ST, RT> ListenableFuture<RT> transformAndFlatten(ListenableFuture<ST> sourceFuture, 
                                                                     Function<? super ST, ListenableFuture<RT>> transformFunction, 
                                                                     Executor executor) {
      return thing(
        sourceFuture, 
        transformFunction.andThen(FutureUtils::immediateResultFuture), // pure . transform
        new FailurePropogatingFutureCallback<ST>(slf) {
          @Override
          public void handleResult(ST result) {
            try {
              slf.setResult(transformFunction.apply(result));
            } catch (Throwable t) {
              slf.setFailure(t);
            }
          }
        });
    }
  }

 protected static <ST, RT> ListenableFuture<RT> transformAndFlatten(ListenableFuture<ST> sourceFuture, 
                                                                     Function<? super ST, ListenableFuture<RT>> transformFunction, 
                                                                     Executor executor) {
      return thing(
        sourceFuture,
        transformFunction,
        new FailurePropogatingFutureCallback<ST>(slf) {
          @Override
          public void handleResult(ST result) {
            try {
              transformFunction.apply(result)
                               .addCallback(new FailurePropogatingFutureCallback<RT>(slf) {
                @Override
                public void handleResult(RT result) {
                  slf.setResult(result);
                }
              });
            } catch (Throwable t) {
              slf.setFailure(t);
            }
          }
        });
    }
  }
```